src/components/Datatable/Pagination/Pagination.tsx

```diff
@@ -0,0 +1,206 @@
+import PageButton from './PageButton';
+import NavButton from './NavButton';
+import { StyledNumber } from '../../Filters/components/Number/Number';
+
+const MAX_PAGE_BUTTONS = 8;
+
+const PaginationWrapper = styled(FlexContainer)`
+  background-color: ${getColor('graphite3H')};
+  padding: ${pxToRem(16)};
+`;
+
+const Ellipsis = styled.div`
+  display: inline-block;
+  width: ${pxToRem(48)};
+  height: ${pxToRem(32)};
+  padding: ${pxToRem(8, 16)};
+  margin: ${pxToRem(0, 4)};
+  color: ${getColor('graphite2B')};
+  background-color: ${getColor('transparent')};
+`;
+
+const PageInputLabel = styled(Label)`
+  margin: ${pxToRem(6, 16)};
+`;
+const PageInput = styled(StyledNumber)`
+  width: ${pxToRem(72)};
+  height: ${pxToRem(32)};
+  padding: ${pxToRem(8, 16)};
+`;
+
+const Pagination: React.FC<PaginationProps> = ({
+  gotoPage,
+  previousPage,
+  nextPage,
+  pageCount,
+  canPreviousPage,
+  canNextPage,
+  pageIndex,
+  isLoading,
+}) => {
+  const pages = Array.from(Array(pageCount), (_, i) => i + 1);
+
+  const [visiblePageButtons, setVisiblePageButtons] = useState(pages);
+  const [isFirstEllipsisVisible, setIsFirstEllipsisVisible] = useState(false);
+  const [isLastEllipsisVisible, setIsLastEllipsisVisible] = useState(false);
+  const [isInputPageInvalid, setIsInputPageInvalid] = useState(false);
+
+  const onChangePageNumber = (e) => {
+    const pageNumber = Number(e.target.value);
+    if (pageNumber <= pageCount && pageNumber > 0) {
+      setIsInputPageInvalid(false);
+      gotoPage(pageNumber - 1);
+    } else {
+      setIsInputPageInvalid(true);
+    }
+  };
+
+  useEffect(() => {
+    if (pageCount > MAX_PAGE_BUTTONS) {
+      if (pageIndex > MAX_PAGE_BUTTONS - 3) {
+        setIsFirstEllipsisVisible(true);
+      }
+      if (pageIndex <= MAX_PAGE_BUTTONS - 3) {
+        setIsFirstEllipsisVisible(false);
+        if (!visiblePageButtons.includes(pageIndex)) {
+          setVisiblePageButtons([2, 3, 4, 5, 6]);
+        }
+      }
+
+      if (pageIndex < pageCount - MAX_PAGE_BUTTONS + 3) {
+        setIsLastEllipsisVisible(true);
+      }
+      if (
+        pageIndex >= MAX_PAGE_BUTTONS - 3 &&
+        pageIndex < pageCount - MAX_PAGE_BUTTONS + 3
+      ) {
+        setIsLastEllipsisVisible(true);
+        if (!visiblePageButtons.includes(pageIndex + 1)) {
+          setVisiblePageButtons([
+            pageIndex,
+            pageIndex + 1,
+            pageIndex + 2,
+            pageIndex + 3,
+          ]);
+        }
+      }
+
+      if (pageIndex >= pageCount - MAX_PAGE_BUTTONS + 2) {
+        setIsLastEllipsisVisible(false);
+        setVisiblePageButtons([
+          pageCount - 5,
+          pageCount - 4,
+          pageCount - 3,
+          pageCount - 2,
+          pageCount - 1,
+        ]);
+      }
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [pageIndex, pageCount]);
+
+  return (
+    <PaginationWrapper>
+      {isLoading && (
```

Comment:
we should give the loading dedicated space even if it is not shown, because the page numbers jumping around when the spinner is visible.